refactor(Shared_url_query): Fix shared  query URL access for SQL Lab users.

LevisNgigi opened the pull request:
### SUMMARY
This update fixes the copy sharing to use the newly introduced permalink.

Fixes #31367

fix(sql-lab): fix the copy url to use the permalink api and allow query sharing for users with appropriate permissions.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
![1737369289283](https://github.com/user-attachments/assets/a4c0ec56-6642-48c8-b024-9d0b3e372e63)


### TESTING INSTRUCTIONS
1. Log in and save a query in SQL Lab.
2. Copy the query URL using the "Copy Query URL" option.
3. Share the URL with a different user (one with the `sqllab` role and access to the dataset).
4. Have the other user try to load the shared query URL.
5. Observe that the second user with sqllab_role can successfully load the shared query.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
- [x] Has associated issue: Fixes #31367
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

LevisNgigi committed:
(refactor)Fix saved query access for SQL Lab users

villebro commented:
Tagging @yousoph @geido as you were active on the original issue. Has there has been discussion about making queries shareable via permalinks? Was this considered as a solution? As @rusackas commented, exposing saved queries to anyone with a link to it can be problematic, as per the current design they are private objects. Implementing this is of course more work, but is aligned with how sharing is done in Explore and the Dashboard views.

LevisNgigi commented:
> Tagging @yousoph @geido as you were active on the original issue. Has there has been discussion about making queries shareable via permalinks? Was this considered as a solution? As @rusackas commented, exposing saved queries to anyone with a link to it can be problematic, as per the current design they are private objects. Implementing this is of course more work, but is aligned with how sharing is done in Explore and the Dashboard views.

Hey, just a clarification.Not anyone with the link can view the query.Any user with public, Gamma, or Alpha roles cannot view the saved query link. Access is strictly limited to users who have been explicitly granted permission by the admin under the SQL Lab role. This ensures that only those with appropriate SQL Lab permissions and dataset access are able to load and view the query. All other users, including those with general roles like Gamma or Alpha, will receive an "Access Denied" message when attempting to access the link.

villebro commented:
Thanks for the clarification @LevisNgigi 👍

LevisNgigi commented:
> Thanks for the clarification @LevisNgigi 👍

@villebro Much Welcome.

justinpark commented:
@LevisNgigi Has there has been discussion about making queries shareable via permalinks (using "copy link")?

LevisNgigi commented:
> @LevisNgigi Has there has been discussion about making queries shareable via permalinks (using "copy link")?

Hey @justinpark I do not think so. I'm not aware of any discussion in reference to the above.

LevisNgigi committed:
Fix: Update SavedQuery filter to resolve access issues and failing tests

LevisNgigi committed:
Fix: Update SavedQuery filter to resolve access issues and failing tests

LevisNgigi committed:
Fix: Update SavedQuery filter to resolve access issues and failing tests

LevisNgigi committed:
Fix: Update SavedQuery filter to resolve access issues and failing tests

LevisNgigi committed:
refactor: modify savedquery filter logic

LevisNgigi committed:
refactor: modify savedquery filter logic

LevisNgigi committed:
refactor: modify savedquery filter logic

LevisNgigi committed:
fix:failing tests

LevisNgigi committed:
fix:failing tests

LevisNgigi committed:
Fix:allow sqllab users to share and view queries

LevisNgigi committed:
Fix:allow sqllab users to share and view queries

LevisNgigi committed:
Fix:fix failing unit tests

geido commented:
Hey @LevisNgigi I am a bit confused about the latest changes where it is just showing a newly introduced endpoint. Would you clarify the approach here? Thank you!

LevisNgigi commented:
> Hey @LevisNgigi I am a bit confused about the latest changes where it is just showing a newly introduced endpoint. Would you clarify the approach here? Thank you!

Hey @geido This endpoint already existed in the project "/api/v1/query/{pk}". I added custom implementation on the endpoint to add ownership validation to ensure only the query owner or authorized users can access it.Those with "can_read", "SavedQuery" permission. The custom implementation I added addresses this gap,by ensuring that users with sqllab role can view shared query via the url.

dpgaspar commented on superset/queries/saved_queries/api.py:
```diff
@@ -196,6 +199,102 @@ def pre_add(self, item: SavedQuery) -> None:
+                return self.response_404()
+
+            is_owner = saved_query.user_id == g.user.id
+            has_access = security_manager.can_access("can_read", "SavedQuery")
```
The `can_read` on `SavedQuery` is already guaranteed by the `@protect` decorator, so it seem redundant to check if the saved query belongs to the user, since it will grant access to all queries to all users.
This does not seem like the correct approach, currently SavedQueries are not shared on Superset, if we want to add this functionality, we can share links using a UUID or wait until we have use group entities and expose queries to groups

LevisNgigi replied:
Thank you for the review and feedback.To clarify, the above implementation does not automatically grant all users access to all saved queries. Instead, it ensures that only users with can_read permissions for SavedQuery can view the query(via a shared url)which is a default role for sqllab. This means that access is already restricted, and not all users will have visibility unless they explicitly have the required permissions.The ownership check was included as an additional layer of security, but I understand your concern about redundancy given the @protect decorator. As you rightly pointed out, one potential vulnerability is the predictability of query IDs in URLs (e.g., http://localhost:9000/sqllab?savedQueryId=3), which makes it possible to guess query IDs. This could expose queries to unauthorized users who have can_read permissions but should not have access to specific queries. Implementing UUIDs for query IDs would mitigate this issue and improve security.Should this be left until query sharing is added in Superset?

LevisNgigi replied:
> @LevisNgigi how hard would it be to implement the UUID approach now?
Implementing the UUID approach would require running a migration script to update existing query IDs to UUIDs in the database while maintaining backward compatibility. Additionally, the logic to verify if the user has the sqllab role would need to be integrated to ensure the correct permissions are applied.I think this adds some complexity especially with the migrations. Unless there is a different approach?

legiondean commented:
@LevisNgigi do you know if there's any workaround setup to make the sharing work before this fix released?
Or any chance the fix could be released soon? Thanks!
The issue is pretty annoying when come to team collaboration.

LevisNgigi commented:
> @LevisNgigi do you know if there's any workaround setup to make the sharing work before this fix released? Or any chance the fix could be released soon? Thanks! The issue is pretty annoying when come to team collaboration.

@legiondean Thanks for bringing this up! I am not aware of any workaround as queries were not meant to be shared originally, but I'm actively working on resolving the issue by implementing a UUID-based approach, which will address the predictability of query IDs and make sharing more secure and reliable.Hopefully it will be released and approved soon. I appreciate your patience and understanding while we work on this!

LevisNgigi committed:
Merge remote-tracking branch 'upstream/master' into fix-saved-query-access

LevisNgigi committed:
Merge branch 'apache:master' into fix-saved-query-access

LevisNgigi committed:
fix:use permalink for query sharing

LevisNgigi committed:
fix:use permalink for query sharing

LevisNgigi committed:
fix:failing tests

LevisNgigi committed:
fix:failing tests

LevisNgigi committed:
fix:use permalink for query sharing

LevisNgigi committed:
fix:restore mouseevent import

betodealmeida commented on superset-frontend/src/pages/SavedQueryList/index.tsx:
```diff
@@ -233,16 +232,31 @@ function SavedQueryList({
+        const payload = {
+          dbId: savedQuery.db_id,
+          name: savedQuery.label,
+          schema: savedQuery.schema,
```
We need to store the catalog as well, for databases that support it.
```suggestion
          catalog: savedQuery.catalog,
          schema: savedQuery.schema,
```

LevisNgigi replied:
Hey, so currently the catalog property does not exist on the SavedQueryObject type [https://github.com/apache/superset/blob/master/superset-frontend/src/views/CRUD/types.ts](Here).

betodealmeida replied:
Ok, let me add it before you merge this, we need it.

LevisNgigi replied:
okay sounds good.

betodealmeida replied:
OK, I tested and it you add `catalog` to `superset-frontend/src/views/CRUD/types.ts` it will be populated automatically by the hook, so you just need to add:

```js
catalog: string | null;
```

to `SavedQueryObject`.

Also, it needs to be added to the Schemas and Types as well, because if a database supports multiple catalogs (like BigQuery or Trino) the query with just the schema is not enough information.

betodealmeida reviewed (changes requested):
This looks very clean! A couple things:

1. Can you test this with a database that supports catalogs? You can use Postgres, but you need to enable "Allow changing catalogs" under Advanced -> Other.
2. Can you add a small test for it?

justinpark commented on superset-frontend/src/pages/SavedQueryList/index.tsx:
```diff
@@ -233,16 +232,31 @@ function SavedQueryList({
+          body: JSON.stringify(payload),
         });
+
+        const { key } = response.json;
+        const permalink = `${window.location.origin}/sqllab/p/${key}`;
```
+1

justinpark commented:
LGTM

LevisNgigi committed:
fix:add tests for copy url icon

LevisNgigi committed:
fix:add tests for copy url icon

LevisNgigi committed:
feat:add catalog to the response payload

betodealmeida reviewed (approved):
Awesome! Thanks for making sure it works with catalogs as well!

LevisNgigi committed:
trigger CI

drummerwolli commented:
@sadpandajoe since i see you are [managing](https://lists.apache.org/thread/m659odpm4mmgsnjrnzvc6c5hwy7qz3pf) the release of 4.1.3, any chance to include this fix still there or in 4.1.4? Here in the PR and in the linked issue, this sounds very severe. And we have in our organization with ca. 2000 users now also this Problem after upgrading to 4.1.1 just last week. I dont think this is a feature release (as someone wrote in the issue) but rather a pretty hefty bug. happy to hear your thoughts.

sadpandajoe commented:
> @sadpandajoe since i see you are [managing](https://lists.apache.org/thread/m659odpm4mmgsnjrnzvc6c5hwy7qz3pf) the release of 4.1.3, any chance to include this fix still there or in 4.1.4? Here in the PR and in the linked issue, this sounds very severe. And we have in our organization with ca. 2000 users now also this Problem after upgrading to 4.1.1 just last week. I dont think this is a feature release (as someone wrote in the issue) but rather a pretty hefty bug. happy to hear your thoughts.

@drummerwolli we don't normally throw refactors into official releases as there could be regression bugs coming out of it. Have you checked out the 5.0rc? This should be part of that official release
